tests/test_multiasaswapper.py

```diff
@@ -0,0 +1,428 @@
+    print(f"\n --- Incentive Wallet {incentive_account.public_key} funded.")
+    return incentive_account
+
+
+@pytest.fixture()
+def offered_asa_a_idx(swap_creator: Wallet) -> int:
+    return mint_asa(
+        swap_creator.public_key,
+        swap_creator.private_key,
+        asset_name="Card A",
+        total=1,
+        decimals=0,
+    )
+
+
+@pytest.fixture()
+def offered_asa_b_idx(swap_creator: Wallet) -> int:
+    return mint_asa(
+        swap_creator.public_key,
+        swap_creator.private_key,
+        asset_name="Card B",
+        total=1,
+        decimals=0,
+    )
+
+
+@pytest.fixture()
+def requested_asa_idx(swap_user: Wallet) -> int:
+    return mint_asa(
+        swap_user.public_key,
+        swap_user.private_key,
+        asset_name="Card B",
+        total=1,
+        decimals=0,
+    )
+
+
+@pytest.fixture()
+def other_asa_idx(swap_user: Wallet) -> int:
+    return mint_asa(
+        swap_user.public_key,
+        swap_user.private_key,
+        asset_name="Other ASA",
+        total=1,
+        decimals=0,
+    )
+
+
+REQUESTED_ALGO_AMOUNT = 1
+
+
+@pytest.fixture()
+def multi_asa_swapper_account(
+    swap_creator: Wallet, offered_asa_a_idx: int, offered_asa_b_idx: int
+) -> LogicSigWallet:
+
+    cfg = MultiAsaSwapConfig(
+        swap_creator=swap_creator.public_key,
+        offered_asa_amounts={
+            str(offered_asa_a_idx): 1,
+            str(offered_asa_b_idx): 1,
+        },
+        requested_algo_amount=1_000_000,
+        max_fee=1_000,
+        optin_funding_amount=BASE_OPTIN_FUNDING_AMOUNT * 2,
+    )
+
+    swapper_lsig = logic_signature(compile_stateless(multi_asa_swapper(cfg)))
+
+    return LogicSigWallet(logicsig=swapper_lsig, public_key=swapper_lsig.address())
+
+
+def test_multi_asa_optin(
+    multi_asa_swapper_account: LogicSigWallet,
+    swap_creator: Wallet,
+    swap_user: Wallet,
+    offered_asa_a_idx: int,
+    offered_asa_b_idx: int,
+    other_asa_idx: int,
+):
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Opt-In fails if not executed by swap creator")
+        swapper_multi_opt_in(
+            swap_creator=swap_user,
+            swapper_account=multi_asa_swapper_account,
+            assets={offered_asa_a_idx: 0, offered_asa_b_idx: 0},
+            funding_amount=BASE_OPTIN_FUNDING_AMOUNT * 2,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Opt-In fails with wrong funding amount")
+        swapper_multi_opt_in(
+            swap_creator=swap_creator,
+            swapper_account=multi_asa_swapper_account,
+            assets={offered_asa_a_idx: 0, offered_asa_b_idx: 0},
+            funding_amount=BASE_OPTIN_FUNDING_AMOUNT - 2,
+        )
+
+    with pytest.raises(AlgodHTTPError):
+        print("\n --- Opt-In fails with wrong ASA")
+        swapper_multi_opt_in(
+            swap_creator=swap_creator,
+            swapper_account=multi_asa_swapper_account,
+            assets={other_asa_idx: 0},
```

Comment:
Should be:

```python
assets={other_asa_idx: 0, other_asa_idx: 0},
```

Reply:
@cusma why does the same ASA repeat twice though? Could you expand on the suggestion?

Reply:
@cusma got it! Just pushed the tweak, resolving